### PR TITLE
HDDS-1430. NPE if secure ozone if KMS uri is not defined. Contributed by Ajay Kumar.

### PR DESCRIPTION
**Pull request by ajayydv**

### CODE REVIEW

File: hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java [hunk 1]
```diff
@@ -59,7 +59,13 @@ public URI getKeyProviderUri() throws IOException {
...
       throws IOException {
-    KeyProvider keyProvider = getKeyProvider();
+    KeyProvider keyProvider;
+    try {
+      keyProvider = getKeyProvider();
+    } catch (IOException ioe) {
+      LOG.error("Error retrieving KeyProvider.", ioe);
+      return null;
+    }
     if (keyProvider instanceof DelegationTokenIssuer) {
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/OzoneFileSystem.java]: Why don't we just check keyProvider!=null on Line 69 before check if keyProvider instanceof? This way, we don't need other changes. 

Note getKeyProvider() will be eventually fixed when OM returns its KMS uri to the client as part of it service discovery.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 21 | Maven dependency ordering for branch |
| +1 | mvninstall | 1126 | trunk passed |
| -1 | compile | 47 | hadoop-ozone in trunk failed. |
| +1 | checkstyle | 40 | trunk passed |
| +1 | mvnsite | 56 | trunk passed |
| +1 | shadedclient | 821 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 71 | trunk passed |
| +1 | javadoc | 40 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 11 | Maven dependency ordering for patch |
| +1 | mvninstall | 54 | the patch passed |
| -1 | compile | 40 | hadoop-ozone in the patch failed. |
| -1 | javac | 40 | hadoop-ozone in the patch failed. |
| +1 | checkstyle | 19 | the patch passed |
| +1 | mvnsite | 46 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 807 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 81 | the patch passed |
| +1 | javadoc | 35 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 24 | client in the patch passed. |
| +1 | unit | 117 | ozonefs in the patch passed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 3569 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/752 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux 06b13754c297 4.4.0-143-generic #169~14.04.2-Ubuntu SMP Wed Feb 13 15:00:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / df76cdc |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/artifact/out/branch-compile-hadoop-ozone.txt |
| findbugs | v3.1.0-RC1 |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/artifact/out/patch-compile-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/testReport/ |
| Max. process+thread count | 3126 (vs. ulimit of 5500) |
| modules | C: hadoop-ozone/client hadoop-ozone/ozonefs U: hadoop-ozone |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-752/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/OzoneKMSUtil.java [hunk 2]
```diff
@@ -128,6 +128,9 @@ public static URI getKeyProviderUri(UserGroupInformation ugi,
 
   public static KeyProvider getKeyProvider(final Configuration conf,
       final URI serverProviderUri) throws IOException{
+    if (serverProviderUri == null) {
```
Comment (xiaoyuyao) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/OzoneKMSUtil.java]: We should not call getKeyProvider when provider uri is not defined. In RpcClient#getKeyProvider() and RestClient#getkeyProvider(), we should do

String kpUri = getKeyProviderUri();
return kpUri == null ? null : OzoneKMSUtil.getKeyProvider(conf, kpUri);

**Comment by xiaoyuyao** [comment]
+1

**Pull request merged**